sdk/core/azure-core/inc/azure/core/azure_assert.hpp

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Provide assert macros to use with pre-conditions.
+ *
+ * @remark Asserts are turned ON when `NDEBUG` is NOT defined (for Debug build). For Release build,
+ * `std::abort()` is directly called if the condition is false, whitout calling assert().
+ *
+ */
+
+#pragma once
+
+#include "platform.hpp"
+
+#include <cstdlib>
+#include <string>
+
+#if defined(NDEBUG)
```

Comment:
I really don't like these macros. 

It doesn't provide a way to disable precondition/postcondition checks. It says in guideline that

```
✔️ YOU MAY provide a means to disable precondition checks in release / optimized builds.
✅ DO provide a way to disable postcondition checks, and omit checking code from built binaries.
```

What kind of problem are we trying to solve in this PR? What are the advantages over just `assert` and `std::abort`?

I can use `assert` for precondition and postcondition, and disable them by `NDEBUG`.
I can use `std::abort` for those checks that are valid in both release and debug build.

Reply:
> What kind of problem are we trying to solve in this PR?

We don't want to disable pre conditions. The guideline needs to be changed. Even on Release, we want to abort if precondition fail.

The advantage is that end-users can update this file (if building from source) to change the behavior

Reply:
> The advantage is that end-users can update this file (if building from source) to change the behavior

I don't know if this is valuable to customers.

Customers are vulnerable to some bugs if they disable assertions. I don't think this is a common scenario in the first place.

Even if they want to, they do it because they hit an assertion failure. In this case, it'd be better if they just disable the one they hit, rather than all assertions.

Reply:
> Customers are vulnerable to some bugs if they disable assertions.

That's why by default we are keeping it for Debug and Release.  On Debug, people uses assert while on Release, abort() is called directly

Reply:
The only difference is in Debug mode, we print some message, while in Release mode, we don't.
The message doesn't hurt anybody even in release mode right?
So why don't we use the same definition regardless of Debug/Release mode?

Reply:
> It doesn't provide a way to disable precondition/postcondition checks.

Actually, it does - albeit only in an awkward way. This:
```c++
#undef NDEBUG
#include <azure/core/azure_assert.hpp>
#define NDEBUG
#include <cassert>

int main() {
  AZURE_ASSERT(false);
}
```
will compile and run without firing the assert.